### DISCUSSION TITLE
Run test cases returned by IXunitTestCaseDiscoverer in parallel

### DISCUSSION THREAD

**Discussion by hach-que** [xunit/xunit#2745] [Question]
I have an `IXunitTestCaseDiscoverer` implementation which returns test cases based on files that exist on the filesystem, and an attribute which associates this discoverer via `[XunitTestCaseDiscoverer]`.

There's only one test method in the test project, because the execution code is the same for all the files on disk. Thus it looks something like this:

```csharp
public class Functional
{
    [Functional]
    public async Task Test(FunctionalTestEntry test)
    {
        // ...
    }
}
```

The problem is that the test cases don't run in parallel, because the collection is via the `Functional` class itself. There doesn't appear to be a way to set the collection per test result returned in the `IXunitTestCaseDiscoverer`.

I also tried implementing `IXunitTestCollectionFactory`, but the function signature doesn't provide the granularity to set a different collection per generated test case:

```csharp
// No way to return different test collections per test case...
ITestCollection Get(ITypeInfo testClass)
```

How can I get test cases generated by `IXunitTestCaseDiscoverer` to run in parallel?

**Comment by bradwilson** [comment]
At the moment (in v2), this isn't easy.

It would require extending a significant portion of the "execution pipeline", and while I do have an example of that, it isn't designed around this. I'm going to put it on my list to come up with a "highly parallel" example because it's a frequent one people have been asking for, and it lets me prototype how to make this much easier in v3.
